Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm

```diff
@@ -63,24 +63,28 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
     return userStyleSheet->userStyleSheet().source() == styleSheetContent.first && userStyleSheet->userStyleSheet().injectedFrames() == injectedFrames;
 }
 
-Vector<RetainPtr<_WKFrameTreeNode>> getFrames(_WKFrameTreeNode *currentNode, std::optional<Vector<WebExtensionFrameIdentifier>> frameIDs)
+static NSArray *getFrames(_WKFrameTreeNode *currentNode, const WebExtensionScriptInjectionParameters& parameters)
 {
-    Vector<RetainPtr<_WKFrameTreeNode>> matchingFrames;
-    Vector<RetainPtr<_WKFrameTreeNode>> framesToCheck { currentNode };
+    NSMutableArray *matchingFrames = [[NSMutableArray alloc] init];
+    Deque<RetainPtr<_WKFrameTreeNode>> framesToCheck { currentNode };
+
+    auto& frameIDs = parameters.frameIdentifiers;
+    auto& documentIDs = parameters.documentIdentifiers;
 
     while (!framesToCheck.isEmpty()) {
-        _WKFrameTreeNode *frame = framesToCheck.first().get();
-        framesToCheck.removeFirst(frame);
+        auto *frame = framesToCheck.takeFirst().get();
 
         auto currentFrameID = toWebExtensionFrameIdentifier(frame.info);
-        if (!frameIDs || frameIDs->contains(currentFrameID))
-            matchingFrames.append(frame);
+        auto currentDocumentID = WTF::UUID::fromNSUUID(frame.info._documentIdentifier);
+
+        if ((!frameIDs && !documentIDs) || (frameIDs && frameIDs->contains(currentFrameID)) || (documentIDs && documentIDs->contains(currentDocumentID)))
+            [matchingFrames addObject:frame];
 
         for (_WKFrameTreeNode *child in frame.childFrames)
             framesToCheck.append(child);
```

Comment:
This is a cool way to avoid recursion. I might have to adopt this in some code I've written lately.